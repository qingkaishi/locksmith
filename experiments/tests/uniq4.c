// Tests that deref of a unique variable does not change its uniqueness.
#include <pthread.h>
#include <stdio.h>

int *j;
pthread_mutex_t l;

void * malloc(int);

void * dostuff(void * a) {
  int *k;
  
  k = (int*) malloc(sizeof (int));
  *k = 2;
  j = k; // race on j, not k
  pthread_mutex_lock(&l);
  j = 3;
  pthread_mutex_unlock(&l);
  return NULL;
}

int main() {
  pthread_t t1;
  pthread_mutex_init(&l, NULL);
  int i = 1;
  pthread_create(&t1, NULL, dostuff, NULL);
  pthread_create(&t1, NULL, dostuff, NULL);
}


