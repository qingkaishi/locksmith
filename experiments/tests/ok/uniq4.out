************************* STARTING *************************


uniq4.c:17: Warning: assigning number to pointer
shared: &j:uniq4.c:5

Warning: Possible data race: &j:uniq4.c:5 is not protected!
 references:
  dereference of &j:uniq4.c:5 at uniq4.c:15
    &j:uniq4.c:5
  locks acquired:
    <empty>
  in: main at uniq4.c:22 

  dereference of &j:uniq4.c:5 at uniq4.c:17
    &j:uniq4.c:5
  locks acquired:
    concrete l:uniq4.c:24
    l:uniq4.c:6
  in: main at uniq4.c:22 


*************************** DONE ***************************
